packages/components/src/unit-control/utils.js

```diff
@@ -281,3 +282,45 @@ export const useCustomUnits = ( { units, availableUnits, defaultValues } ) => {
 
 	return usedUnits.length === 0 ? false : usedUnits;
 };
+
+/**
+ * Get available units with the unit for the currently selected value
+ * prepended if it is not available in the list of units.
+ *
+ * This is useful to ensure that the current value's unit is always
+ * accurately displayed in the UI, even if the intention is to hide
+ * the availability of that unit.
+ *
+ * @param {number|string} currentValue Selected value to parse.
+ * @param {string}        legacyUnit   Legacy unit value, if currentValue needs it appended.
+ * @param {Array<Object>} units        List of available units.
```

Comment:
I've added in a `@typedef` at the top of the file to create a definition of the unit object. Let me know if it needs further changes — I see there's a proposal to add in support for the keyword `any` for the `step` value in https://github.com/WordPress/gutenberg/pull/34542. Since that isn't supported yet, I haven't added it to the type, but happy to add it in if you think we should get it in earlier.